flink-runtime/src/main/java/org/apache/flink/runtime/management/JMXServer.java

```diff
@@ -0,0 +1,239 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.management;
+
+import org.apache.flink.configuration.JMXServerOptions;
+import org.apache.flink.util.NetUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.management.remote.JMXConnectorServer;
+import javax.management.remote.JMXServiceURL;
+import javax.management.remote.rmi.RMIConnectorServer;
+import javax.management.remote.rmi.RMIJRMPServerImpl;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.net.MalformedURLException;
+import java.rmi.NoSuchObjectException;
+import java.rmi.NotBoundException;
+import java.rmi.Remote;
+import java.rmi.RemoteException;
+import java.rmi.registry.Registry;
+import java.rmi.server.UnicastRemoteObject;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+/**
+ * JMX Server implementation that JMX clients can connect to.
+ *
+ * <p>Heavily based on j256 simplejmx project
+ *
+ * <p>https://github.com/j256/simplejmx/blob/master/src/main/java/com/j256/simplejmx/server/JmxServer.java
+ */
+public class JMXServer {
+	private static final Logger LOG = LoggerFactory.getLogger(JMXServer.class);
+
+	private static JMXServer instance = null;
+
+	private final AtomicReference<Remote> rmiServerReference = new AtomicReference<>();
+
+	private Registry rmiRegistry;
+	private JMXConnectorServer connector;
+	private int port;
+
+	/**
+	 * Construct a new JMV-wide JMX server or acquire existing JMX server.
+	 *
+	 * <p>If JMXServer static instance is already constructed, it will not be
+	 * reconstruct again. Instead a warning sign will be posted if the desired
+	 * port configuration doesn't match the existing JMXServer static instance.
+	 *
+	 * @param portsConfig port configuration of the JMX server.
+	 * @return JMXServer static instance.
+	 */
+	public static JMXServer startInstance(String portsConfig) {
+		if (instance == null) {
+			if (!portsConfig.equals(JMXServerOptions.JMX_SERVER_PORT.defaultValue())) {
+				instance = startJMXServerWithPortRanges(portsConfig);
+			} else {
+				LOG.warn("JMX Server start failed. No explicit JMX port is configured.");
+				instance = null;
+			}
+		} else {
+			LOG.warn("JVM-wide JMXServer already started at port: " + instance.port);
+		}
+		return instance;
+	}
+
+	/**
+	 * Acquire existing JMX server. or null if not started.
+	 *
+	 * @return the JMXServer static instance.
+	 */
+	public static JMXServer getInstance() {
+		return instance;
+	}
+
+	/**
+	 * Stop the JMX server.
+	 */
+	public static void stopInstance() throws IOException {
+		if (instance != null) {
+			instance.stop();
+			instance = null;
+		}
+	}
+
+	public static int getPort() {
```

Comment:
return an Optional<Integer> instead, then you can probably also get rid of `JMXServer#getInstance`